Hw5 popov

iliapopov17 opened the pull request.

iliapopov17 committed:
Create betafold-3_fastq_tool.py

iliapopov17 committed:
Add gc_check() function

This function checks if the GC content of a DNA sequence in fastq file lies within specified bounds

Parameters:
- seq (str): DNA sequence
- gc_bounds (tuple): tuple with bottom and top bounds for GC content

Returns:
- bool: True if GC content is within bounds, False if not

iliapopov17 committed:
Add length_check() function

This function checks if the length of a DNA sequence in fastq file lies within specified bounds

Parameters:
- seq (str): DNA sequence
- length_bounds (tuple): tuple with bottom and top bounds for sequence length

Returns:
- bool: True if the length is within bounds, False if not

iliapopov17 committed:
Add quality_check() function

This function checks if the quality score of a DNA sequence in fastq file lies above a specified bound.

Parameters:
- seq (str): DNA sequence
- quality_threshold (int): acceptable quality

Returns:
- bool: True if the average quality score is above the threshold, False if not

iliapopov17 committed:
Add fastq_tool() main function

This is the main function that filters the DNA sequences in a fastq file based on specified criteria

Parameters:
- seqs (dict): dictionary where keys are sequence identifiers and values are tuples containing sequence and quality scores
- gc_bounds (tuple): tuple with bottom and top bounds for GC content
- length_bounds (tuple): tuple with bottom and top bounds for sequence length
- quality_threshold (int): acceptable quality

 Returns:
- dict: filtered dictionary that contains only the sequences that fully match the specified criteria

iliapopov17 committed:
Create betafold-3_amino_analyzer.py

iliapopov17 committed:
Major fixes to betafold-3_amino_analyzer tool

iliapopov17 committed:
Create betafold-3_dna_rna_tools.py

iliapopov17 committed:
Major fixes to betafold-3_dna_rna_tools.py

iliapopov17 committed:
All script names rename

Renamed all script files from betafold-3... to betafold_3...

iliapopov17 committed:
Add import of main functions to main script

iliapopov17 committed:
Rename of main function

Renamed fastq_tool() to run_fastq_tool() in order to keep the code consistent

iliapopov17 committed:
Update README.md

albidgy commented on src/betafold_3_amino_analyzer.py:
```diff
@@ -0,0 +1,203 @@
+AA_SET = set(['V', 'I', 'L', 'E', 'Q', 'D', 'N', 'H', 'W', 'F', 'Y', 'R', 'K', 'S', 'T', 'M', 'A', 'G', 'P', 'C',
+               'v', 'i', 'l', 'e', 'q', 'd', 'n', 'h', 'w', 'f', 'y', 'r', 'k', 's', 't', 'm', 'a', 'g', 'p', 'c'])
```
Зачем список, который ты сам задал и знаешь, что он состоит из уникальных элементов, превращать в множество? Можно сразу сделать set, если ты хочешь с ним работать
```suggestion
AA_SET = {'V', 'I', 'L', 'E', 'Q', 'D', 'N', 'H', 'W', 'F', 'Y', 'R', 'K', 'S', 'T', 'M', 'A', 'G', 'P', 'C',
               'v', 'i', 'l', 'e', 'q', 'd', 'n', 'h', 'w', 'f', 'y', 'r', 'k', 's', 't', 'm', 'a', 'g', 'p', 'c'}
```

albidgy commented on src/betafold_3_amino_analyzer.py:
```diff
@@ -0,0 +1,203 @@
+    - dict: dictionary mapping amino acids to their weights based on the chosen type.
+    """
+    if weight == 'average':
+        average_weights = {
+            'A': 71.0788, 'R': 156.1875, 'N': 114.1038, 'D': 115.0886, 'C': 103.1388,
+            'E': 129.1155, 'Q': 128.1307, 'G': 57.0519, 'H': 137.1411, 'I': 113.1594,
+            'L': 113.1594, 'K': 128.1741, 'M': 131.1926, 'F': 147.1766, 'P': 97.1167,
+            'S': 87.0782, 'T': 101.1051, 'W': 186.2132, 'Y': 163.1760, 'V': 99.1326
+        }
```
Я отдельно это не выделяла в прошлом дз, поэтому баллы за это не сниму, но на будущее: если ты создаешь какой-то список/словарь/множество и тд, куда записываешь какие-то элементы/значения и знаешь, что эта переменная не будет изменяться, то такие переменные лучше делать константами и выносить за пределы функций.

albidgy commented on src/betafold_3_amino_analyzer.py:
```diff
@@ -0,0 +1,203 @@
+    
+    if enzyme == "trypsin":  # Trypsin cuts peptide chains mainly at the carboxyl side of the amino acids lysine or arginine.
+        for aa in range(len(sequence)-1):
+            if sequence[aa] in ['K', 'R', 'k', 'r'] and sequence[aa+1] not in ['P','p']:
```
Пробел
```suggestion
            if sequence[aa] in ['K', 'R', 'k', 'r'] and sequence[aa+1] not in ['P', 'p']:
```

albidgy commented on src/betafold_3_amino_analyzer.py:
```diff
@@ -0,0 +1,203 @@
+    
+    if enzyme == "chymotrypsin":  # Chymotrypsin preferentially cleaves at Trp, Tyr and Phe in position P1(high specificity) 
+        for aa in range(len(sequence) - 1):
+            if sequence[aa] in ['W', 'Y', 'F', 'w', 'y', 'f'] and sequence[aa+1] not in ['P','p']:
```
```suggestion
            if sequence[aa] in ['W', 'Y', 'F', 'w', 'y', 'f'] and sequence[aa + 1] not in ['P','p']:
```

albidgy commented on betafold_3.py:
```diff
@@ -0,0 +1,3 @@
+from src.betafold_3_dna_rna_tools import run_dna_rna_tools
+from src.betafold_3_amino_analyzer import run_amino_analyzer
+from src.betafold_3_fastq_tool import run_fastq_tool
```
Согласно заданию предполагалось, что в этом скрипте должны лежать 3 главные функции.
<img width="501" alt="Снимок экрана 2023-10-12 в 15 05 30" src="https://github.com/iliapopov17/BetaFold-3/assets/90711855/cf3440c6-322e-42d9-ad88-91b6e0507464">

albidgy commented on src/betafold_3_amino_analyzer.py:
```diff
@@ -0,0 +1,203 @@
+    if enzyme not in ("trypsin", "chymotrypsin"):
+        return "You have chosen an enzyme that is not provided. Please choose between trypsin and chymotrypsin."
+    
+    if enzyme == "trypsin":  # Trypsin cuts peptide chains mainly at the carboxyl side of the amino acids lysine or arginine.
```
Просто на будущее: обычно используют или одинарные, или двойные кавычки. Лучше выбрать один стиль.

albidgy commented on src/betafold_3_dna_rna_tools.py:
```diff
@@ -0,0 +1,119 @@
+    - str: transcribed RNA sequence
+    """
+
+    rna_sequence = ''
```
Вообще лучше не использовать конкатенацию строк, а добавлять элементы в список

albidgy commented on src/betafold_3_dna_rna_tools.py:
```diff
@@ -0,0 +1,119 @@
+    - str: complemented sequence
+    """
+
+    complement_DNA = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C',
+                       'a': 't', 't': 'a', 'c': 'g', 'g': 'c'}
+    complemetn_RNA = {'A': 'U', 'G': 'C', 'U': 'A', 'C': 'G',
+                      'a': 'u', 'g': 'c', 'u': 'a', 'c': 'g',}
```
Антон это не писал отдельным комментарием, но лучше сделать константами

albidgy commented on src/betafold_3_fastq_tool.py:
```diff
@@ -0,0 +1,121 @@
+
+    for base in sequence:
+        if base == "G" or base == "C":
+            gc += 1
```
А где деление на длину? В задании же не в штуках нужно было посчитать GC состав, а в процентах
<img width="415" alt="Снимок экрана 2023-10-12 в 15 31 38" src="https://github.com/iliapopov17/BetaFold-3/assets/90711855/32f5e3d3-af24-44a8-815c-39f69ece4f08">

albidgy commented on src/betafold_3_fastq_tool.py:
```diff
@@ -0,0 +1,121 @@
+    for name, (sequence, quality) in seqs.items():        
+        if ((gc_check(sequence, gc_bounds))
+            and (length_check(sequence, length_bounds))
+            and (quality_check(quality, quality_threshold))):
```
Скобки не нужны
```suggestion
        if (gc_check(sequence, gc_bounds) 
            and length_check(sequence, length_bounds) 
            and quality_check(quality, quality_threshold)):
```

albidgy commented on src/betafold_3_fastq_tool.py:
```diff
@@ -0,0 +1,121 @@
+    if type(gc_bounds) == int:
+        gc_bounds = (0, gc_bounds)
+
+    elif len(gc_bounds) == 1:
```
Не самая удачная идея. Если подать число с плавающей точкой, код работать не будет.
<img width="1082" alt="Снимок экрана 2023-10-12 в 15 38 25" src="https://github.com/iliapopov17/BetaFold-3/assets/90711855/1456a021-45a1-47fb-9d3c-7da3799d3cbf">

Здесь лучше было делать проверку не на длину gc_bounds, а сделать дополнительную проверку на тип float

albidgy commented on src/betafold_3_fastq_tool.py:
```diff
@@ -0,0 +1,121 @@
+    gc = 0
+
+    for base in sequence:
+        if base == "G" or base == "C":
```
Не считывается регистр

albidgy commented on src/betafold_3_fastq_tool.py:
```diff
@@ -0,0 +1,121 @@
+    if type(length_bounds) == int:
+        length_bounds = (0, length_bounds)
+
+    elif len(length_bounds) == 1:
+        length_bounds = (0, length_bounds[0])
```
Эта проверка ничего не даст. В принципе не логично подавать не целое число или интервал. Но тут если подать float будет та же самая ошибка

albidgy commented on src/betafold_3_fastq_tool.py:
```diff
@@ -0,0 +1,121 @@
+    elif len(length_bounds) == 1:
+        length_bounds = (0, length_bounds[0])
+
+    elif len(length_bounds) > 2:
+        raise ValueError("Invalid length_bounds value!")
+
+    elif type(length_bounds) != tuple:
+        raise ValueError("Input tuple value!")
+
+    if len(str(quality_threshold)) > 2:
+        raise ValueError("Invalid quality_threshold value!")
+    
+    elif type(quality_threshold) != int:
+        raise ValueError("Input integer value!")
```
Все эти проверки лучше было вынести в отдельную функцию

albidgy commented on src/betafold_3_fastq_tool.py:
```diff
@@ -0,0 +1,121 @@
+    elif type(gc_bounds) != tuple:
+        raise ValueError("Input tuple value!")
+
+    if type(length_bounds) == int:
```
вообще лучше использовать isinstance
```suggestion
    if isinstance(length_bounds, int):
```

albidgy commented on src/betafold_3_dna_rna_tools.py:
```diff
@@ -0,0 +1,119 @@
+
+    if len(results) == 1:
+        return results[0]
+    else:
+        return results
```
Антон выделял это, как улучшение
![Uploading Снимок экрана 2023-10-12 в 16.09.42.png…]()

albidgy commented on src/betafold_3_amino_analyzer.py:
```diff
@@ -0,0 +1,203 @@
+    
+    if cleavage_sites:
+        print( f"Found {len(cleavage_sites)} {enzyme} cleavage sites at positions {', '.join(map(str, cleavage_sites))}")
+        return cleavage_sites#  list(', '.join(map(str, cleavage_sites)))
```
Странный комментарий. Видимо, это закомментированная строка, такое лучше вычищать.

albidgy reviewed (commented):
__Плюсы:__
- Комментарии к коммитам осознанные, мне нравятся
- Здорово, что README сделан в едином стиле
- Стиль кода стал значительно лучше!

__Замечания:__
- функция на проверку gc_bound работает идейно неверно. Нужно было посчитать процент GC в последовательности. 
- В главной функции по работе с FASTQ нельзя передать передать float. Здесь же есть 2 бессмысленные проверки на длину = 1 (см. комментарии).
- Обрати внимание, на константы.
- Неправильно сделан главный скрипт, там должно было лежать 3 главных функции, поэтому импорты стали бесполезными.
- Не нужно конкатенировать строки, лучше использовать списки.

__Баллы:__
- 3 фильтрации FASTQ 2.2/3 (gc неправильно считается, не учитывает регистр)
- Главная функция 0.7/1 (нельзя подать float в gc_bounds)
- README 2/2
- Структура репозитория и качество кода 1.8/3. -0.2 за небольшие неточности в стиле кода (например, ненужные скобки в if, отсутствие пробелов). -1 за неправильно сделанный скрипт betafold_3.py.
- Улучшение кода ДНК/РНК и белковых тулов 0.5/1

__Итого:__ 7.2 балла

iliapopov17 committed:
Fix gc_check() function

The function now checks the percentage of GC content

iliapopov17 committed:
Merge branch 'hw5_popov' of https://github.com/iliapopov17/BetaFold-3 into hw5_popov

iliapopov17 committed:
Update src/betafold_3_amino_analyzer.py

Fix: Зачем список, который ты сам задал и знаешь, что он состоит из уникальных элементов, превращать в множество? Можно сразу сделать set, если ты хочешь с ним работать

Co-authored-by: Alexandra Kasianova <90711855+albidgy@users.noreply.github.com>

iliapopov17 committed:
Update src/betafold_3_amino_analyzer.py

Add space

Co-authored-by: Alexandra Kasianova <90711855+albidgy@users.noreply.github.com>

iliapopov17 committed:
Update src/betafold_3_amino_analyzer.py

Add space

Co-authored-by: Alexandra Kasianova <90711855+albidgy@users.noreply.github.com>

iliapopov17 committed:
Update src/betafold_3_fastq_tool.py

Remove unnecessary brackets

Co-authored-by: Alexandra Kasianova <90711855+albidgy@users.noreply.github.com>

iliapopov17 committed:
Update README.md
